test/common_nn.py

```diff
@@ -769,13 +779,23 @@ def test_cuda(self, test_case):
             cpu_module = self.constructor(*self.constructor_args)
             gpu_module = self.constructor(*self.constructor_args).float().cuda()
 
-            cpu_output = test_case._forward_criterion(cpu_module, cpu_input, cpu_target)
-            gpu_output = test_case._forward_criterion(gpu_module, gpu_input, gpu_target)
+            reduce = get_reduce(cpu_module)
+            cpu_output = test_case._forward_criterion(
+                cpu_module, cpu_input, cpu_target, reduce=reduce)
+            gpu_output = test_case._forward_criterion(
+                gpu_module, gpu_input, gpu_target, reduce=reduce)
             test_case.assertEqual(cpu_output, gpu_output, 4e-4)
 
-            cpu_gradInput = test_case._backward_criterion(cpu_module, cpu_input, cpu_target)
-            gpu_gradInput = test_case._backward_criterion(gpu_module, gpu_input, gpu_target)
+            cpu_gradOutput = \
+                None if reduce else torch.randn(cpu_input.size()).double()
+            gpu_gradOutput = \
+                None if reduce else to_gpu(cpu_gradOutput, type_map=type_map)
```

Comment:
This is confusing. Why do you give it a `None` grad if `reduce` is true?

Reply:
The legacy cuda test for `MSECriterion` (`TestNN.test_MSECriterion_cuda`) also runs through this code. Right now I'm having the legacy test code [assume that `gradOutput` is `None`](https://github.com/pytorch/pytorch/pull/2878/files?diff=split#diff-b366aa5d4d0d6672aa91c10b561eaaf1R639) because I'm leaving the legacy code as-is and I don't think the legacy code supports having a `gradOutput`. 

I do agree that having `None` here is pretty confusing to read. I could change [the legacy test code](https://github.com/pytorch/pytorch/pull/2878/files?diff=split#diff-b366aa5d4d0d6672aa91c10b561eaaf1R639) to just ignore `gradOutput` instead of asserting that it's `None` -- what do you think about that?